app.go

```diff
@@ -0,0 +1,68 @@
+package tscaddy
+
+import (
+	"errors"
+
+	"github.com/caddyserver/caddy/v2"
+)
+
+func init() {
+	caddy.RegisterModule(TSApp{})
+}
+
+type TSApp struct {
+	// DefaultAuthKey is the default auth key to use for Tailscale if no other auth key is specified.
+	DefaultAuthKey string `json:"auth_key,omitempty" caddy:"namespace=tailscale.auth_key"`
+
+	Ephemeral bool `json:"ephemeral,omitempty" caddy:"namespace=tailscale.ephemeral"`
+
+	Servers map[string]TSServer `json:"servers,omitempty" caddy:"namespace=tailscale"`
+}
+
+type TSServer struct {
+	AuthKey string `json:"auth_key,omitempty" caddy:"namespace=auth_key"`
+
+	Ephemeral bool `json:"ephemeral,omitempty" caddy:"namespace=tailscale.ephemeral"`
+
+	name string
+}
+
+func (TSApp) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  "tailscale",
+		New: func() caddy.Module { return new(TSApp) },
+	}
+}
+
+func (t *TSApp) Provision(ctx caddy.Context) error {
+	for _, svr := range t.Servers {
+		if t.Ephemeral {
+			svr.Ephemeral = t.Ephemeral
+		}
+		if svr.AuthKey == "" {
+			svr.AuthKey = t.DefaultAuthKey
+		}
+	}
+
+	app = t
+
+	return nil
+}
+
+func (t *TSApp) Validate() error {
+	if t.DefaultAuthKey == "" {
+		return errors.New("auth_key must be set")
+	}
+	return nil
+}
+
+func (t *TSApp) Start() error {
+	return nil
+}
+
+func (t *TSApp) Stop() error {
+	app = nil
```

Comment:
If the apps are started and stopped at the same time would we do this at all? 

Could we have a race condition where stop on an old app is called after start on a new one?

Reply:
I had the same thought.  For updating configs this isn't really necessary because the new TSApp's Start will overwrite the value before Stop is called.  But if the new config removes the Tailscale configuration, we still want to clean this up.  The use of atomic and CompareAndSwap handle race conditions.

Reply:
Correct me if I'm wrong but `atomic` will solve concurrent access but not an ordering error where the calls are not guaranteed to always be in the same order?

Reply:
Yes, you're absolutely right that atomic is focused on handling concurrency.  It's the CompareAndSwap that helps us here.  On Start, we always update the value of our singleton to the new value.  But on Stop, we only set it to nil if the value is still the old value.  So depending on the order they run...


``` go
old.Stop() // this will update tsapp to nil
new.Start() // this will update tsapp to new

new.Start() // this will update tsapp to new
old.Stop() // this won't do anything because tsapp no longer points to old, which is what CompareAndSwap checks for
```

In the case of the new caddy config not having a tailscale config at all, then only `old.Stop()` gets called, which will set tsapp to nil and free up the memory.

Reply:
Got it. Thanks for the explanation!